sos/plugins/etcd.py

```diff
@@ -16,21 +16,39 @@
 # 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 
 from sos.plugins import Plugin, RedHatPlugin
+import socket
 
 
 class etcd(Plugin, RedHatPlugin):
 
     """etcd plugin
     """
 
+    def port_report(self):
+        """ Until etcd v2, the daemon listens on port 4001 for backward
+        compatibility of v2.0 eariler. But etcd v3 no loger litens on 
+        port 4001 and it listens on port 2379 only. 
+        To collect information in any version, this plugin should check 
+        which port is available.
+        """
+        self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        try:
+            self.sock.connect(("localhost", 2379))
+        except socket.error, msg:
+            self.sock.close()
+            return "4001"
+        self.sock.close()
+        return "2379"
+
     def setup(self):
         self.add_copy_spec("/etc/etcd")
 
-        self.add_cmd_output("curl http://localhost:4001/version")
-        self.add_cmd_output("curl http://localhost:4001/v2/members")
-        self.add_cmd_output("curl http://localhost:4001/v2/stats/leader")
-        self.add_cmd_output("curl http://localhost:4001/v2/stats/self")
-        self.add_cmd_output("curl http://localhost:4001/v2/stats/store")
+        curl_command = "curl -s http://localhost:" + str(self.port_report())
+        self.add_cmd_output(str(curl_command) + "/version")
+        self.add_cmd_output(str(curl_command) + "/v2/members")
+        self.add_cmd_output(str(curl_command) + "/v2/stats/leader")
+        self.add_cmd_output(str(curl_command) + "/v2/stats/self")
+        self.add_cmd_output(str(curl_command) + "/v2/stats/store")
```

Comment:
If `curl_command` is not a string here (requiring coercion with `str()`) then something is already horribly wrong. It's also wasteful to do it five times over.